cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java

```diff
@@ -387,11 +392,23 @@ private void executeCustomAction(final WorkflowActionNode node, InstantiatorFact
     CustomActionExecutor customActionExecutor = new CustomActionExecutor(workflowRunId, context,
                                                                          instantiator, classLoader);
     status.put(node.getNodeId(), node);
+    store.addWorkflowNodeState(workflowRunId, new WorkflowNodeStateDetail(node.getNodeId(), NodeStatus.RUNNING, null,
+                                                                          null));
+    Throwable failureCause = null;
     try {
       customActionExecutor.execute();
+    } catch (Throwable t) {
+      failureCause = t;
+      Throwables.propagateIfPossible(t, Exception.class);
+      throw Throwables.propagate(t);
     } finally {
       status.remove(node.getNodeId());
       store.updateWorkflowToken(workflowRunId, token);
+      NodeStatus status = failureCause == null ? NodeStatus.COMPLETED : NodeStatus.FAILED;
+      nodeStates.put(node.getNodeId(), new WorkflowNodeState(node.getNodeId(), status, null, failureCause));
+      store.addWorkflowNodeState(workflowRunId, new WorkflowNodeStateDetail(node.getNodeId(), status, null,
+                                                                            failureCause == null ? null
```

Comment:
nit: move the ``Throwable`` definition to a variable to improve the indentation here